src/Donors/Controllers/DonorRequestController.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Give\Donors\Controllers;
+
+use Give\Donations\ValueObjects\DonationMetaKeys;
+use Give\Donors\Models\Donor;
+use Give\Donors\Repositories\DonorRepository;
+use Give\Donors\ValueObjects\DonorRoute;
+use Give\Framework\QueryBuilder\JoinQueryBuilder;
+use Give\Framework\QueryBuilder\QueryBuilder;
+use WP_Error;
+use WP_REST_Request;
+use WP_REST_Response;
+
+/**
+ * @unreleased
+ */
+class DonorRequestController
+{
+    /**
+     * @unreleased
+     *
+     * @return WP_Error | WP_REST_Response
+     */
+    public function getDonor(WP_REST_Request $request)
+    {
+        $donation = Donor::find($request->get_param('id'));
+
+        if ( ! $donation) {
+            return new WP_Error('donor_not_found', __('Donor not found', 'give'), ['status' => 404]);
+        }
+
+        return new WP_REST_Response($donation->toArray());
+    }
+
+    /**
+     * @unreleased
+     */
+    public function getDonors(WP_REST_Request $request): WP_REST_Response
+    {
+        $page = $request->get_param('page');
+        $perPage = $request->get_param('per_page');
+
+        $query = give(DonorRepository::class)->prepareQuery();
+
+        if ($campaignId = $request->get_param('campaignId')) {
+            $query->select(['donationmeta1.donation_id', 'donationId'])
+                ->distinct()
+                ->join(function (JoinQueryBuilder $builder) use ($campaignId) {
+                    $builder->innerJoin('give_donationmeta', 'donationmeta1')
+                        ->joinRaw("ON donationmeta1.meta_key = '" . DonationMetaKeys::DONOR_ID . "' AND donationmeta1.meta_value = ID");
+
+                    $builder->innerJoin('give_donationmeta', 'donationmeta2')
+                        ->joinRaw("ON donationmeta2.meta_key = '" . DonationMetaKeys::CAMPAIGN_ID . "' AND donationmeta2.meta_value = {$campaignId}");
+                })->whereIn('donationmeta1.donation_id', function (QueryBuilder $builder) {
+                    $builder
+                        ->select('ID')
+                        ->from('posts')
+                        ->whereRaw("WHERE ID = donationmeta1.donation_id AND post_type = 'give_payment' AND post_status = 'publish'");
+                });
+        }
```

Comment:
If I understand this corrently, this is querying the Donors table and joining in the Donations associated with a specific Campaign, then limiting those results based on the status of the joined Donation?

Would it make more sense to query the Donors table where in a list of Donations associated with the Campaign? This would encapsulate the donation query, instead of conflating the two.

The end result is the same, but the structure of the relationship is different.

From a maintenance perspective I think this would be easier to understand when returning to this code.

Thinking out loud, the `select` for the `donationId` and the use of `distinct` stood out as possible red flags.

Reply:
We could also leverage the [CampaignDonationQuery](https://github.com/impress-org/give/blob/18049358a9b75ea37a57f04cd4567b18ab0647d5/src/Campaigns/CampaignDonationQuery.php#L14), which would also account for the payment mode and include subscription donations.

```php
(new CampaignDonationQuery($campaign))
    ->joinDonationMeta(DonationMetaKeys::DONOR_ID, 'donorId')
    ->select('donorId')
    ->distinct()
```

Reply:
I fully refactor how this query works...

Since I did a bunch of commits with these changes, I'll not copy and paste all of them here because it will require a new full review in this PR.